[WIP] Server modules

filrak opened the pull request:
### Related issues
<!--  Put related issue number which this PR is closing. For example #123 -->

closes #2494

### Short description and why it's useful
<!-- describe in a few words what is this Pull Request changing and why it's useful -->


todo: ability to pass config to the module in `local.json`

How to use:

1. create `server.js` in the root of your module and export function that will extend the express context:
```js
module.exports = function (app) {
  app.get('/robots.txt', (req, res) => {
    res.end('User-agent: *\nDisallow: ')
  })
}
```
or with config
```js
module.exports = function (app, config) {
  app.get('/robots.txt', (req, res) => {
    res.end(config.content)
  })
}

```
2. Register server side module in `config/local.json`
```
  "serverSideModules": [
    "src/modules/example"
  ],
```
or with config
```
  "serverSideModules": [
    ["src/modules/example", { "content" : "User-agent: *\nDisallow: " }
  ],
```

voila!

**NOTE**: `serverSideModules` property is filtered on client side because it can contain fragile data such as API keys, also it's redundant anyway because it's server side ^^
### Screenshots of visual changes before/after (if there are any)
<!-- if you made any changes in the UI layer please provide before/after screenshots -->

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance
<!-- Please get familiar with following info -->

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

filrak commented:
Imho it would be nice to also filter out serverSideModules field from the config json on client side. It will probably contain fragile info such as API keys

filrak commented:
@kkdg can you add this to modules docs?

pkarw commented on core/scripts/server.js:
```diff
@@ -120,10 +121,18 @@ app.use('/service-worker.js', serve('dist/service-worker.js', false, {
+config.serverSideModules.forEach(serverModule => {
+  if (Array.isArray(serverModule)) {
+    require(resolve(serverModule[0] + '/server.js'))(app, serverModule[1])
+  } else {
```
Cool, but few questions:
- Does it support ES6/TS in server modules? Otherwise no possibility to reuse VS code and libs
- do we really need this config variable? It’s one time operation so we can probably traverse core/modules and src/modules(?)
- we need an example module in the module template

ResuBaka replied:
@pkarw 

I can say for the first point that it does not support typescript, but that would be a really nice feature.

patzick replied:
yeah, would be great to have these modules in separate dir or register in some index.js file like modules, this index.js would always be consumed on the server side. I'm afraid that config option could be overcomplicated.

And example +1 :)

filrak replied:
@pkarw @patzick  imho we need to if you want to include modules from node_modules as well. We can't really assume that their always named `vsf-something`. 
Separate index.js is an optiion as well but for sure we can't just traverse the file tree. Please pick either config or file.

About es6 tbh I don't know, will check it. It works exactly the same as server.js extensions

patzick replied:
so I'm voting for registering in a file instead of config :)

pkarw commented:
@patzick @filrak I get the idea with config - so I vote for config.
But we should have the TypeScript/ES6 support and test it with our Vuex. I mean user should be ably to import store and query elasticsearch (eg. `product/list`) in server module; otherwise it will be pretty much not usefull feature

filrak commented:
Ok i applied suggested . hanges:
- you can now use ES6 is server side modules (i also got rid of redundant `entry.ts` and just changed zserver.js` to typescript file)
- instead of using config there is `modules/server.ts` file, also renamed `index.ts` to `client.ts` for constancy
- I removed `src/server` folder to keep common approach for server side extensions via modules

Please lmk if that's fine, then I will add upgrade notes

kkdg commented:
Sure thing!

kkdg commented:
@filrak This is one of the most anticipated parts of my interest. I will start working on it after 1.10 -> 1.11 upgrade tutorial is done as @pkarw told me to do. thanks!

filrak commented:
sure!
